numpy/_core/src/multiarray/unique.cpp

```diff
@@ -18,77 +24,243 @@ struct FinalAction {
   private:
     F clean_;
 };
-
 template <typename F>
 FinalAction<F> finally(F f) {
     return FinalAction<F>(f);
 }
 
-template<typename T>
+template <typename T>
 static PyObject*
-unique(PyArrayObject *self)
+unique_integer(PyArrayObject *self, npy_bool equal_nan)
 {
-    /* This function takes a numpy array and returns a numpy array containing
-    the unique values.
+    /*
+    * Returns a new NumPy array containing the unique values of the input array of integer.
+    * This function uses hashing to identify uniqueness efficiently.
+    */
+    NPY_ALLOW_C_API_DEF;
+    NPY_ALLOW_C_API;
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    // NumPy API calls and Python object manipulations require holding the GIL.
+    Py_INCREF(descr);
+    NPY_DISABLE_C_API;
+
+    // release the GIL
+    PyThreadState *_save1 = PyEval_SaveThread();
+
+    // number of elements in the input array
+    npy_intp isize = PyArray_SIZE(self);
+
+    // Reserve hashset capacity in advance to minimize reallocations and collisions.
+    // We use min(isize, 1024) as the initial bucket count:
+    // - Reserving for all elements (isize) may over-allocate when there are few unique values.
+    // - Using a moderate upper bound (1024) keeps memory usage reasonable (4 KiB for pointers).
+    // See discussion: https://github.com/numpy/numpy/pull/28767#discussion_r2064267631
+    std::unordered_set<T> hashset(std::min(isize, (npy_intp)1024));
```

Comment:
can you make this an ALL_CAPS_MACRO defined at global scope instead of the literal 1024?